### DISCUSSION TITLE
Forwarding the message to the next node

### DISCUSSION THREAD

**Discussion by polentacremosa** [LoRaMesher/LoRaMesher#55] [Q&A]
Hi everyone,

First of all, I want to thank for the library and the published articles related to it.

I'm new to working with the library and Lora, and I could use some assistance. 
I managed to run the Counter and Display example code on an ESP-32 Wifi Lora V2. But I got an error when the message was sent to a node within 2 hops distance (metric 2).

Could anybody give any guidance on adapting the code to forward messages to the next node? Or  If anyone has an example code for this implementation.

Thank you!

**Comment by Jaimi5** [comment]
Hi @danieltsutsumi 
What was the error you got? The library should forward the packets automatically if you send them to the node.
Can you give me some logs to try to reproduce the error?

Thank you for using the library!

**Reply by polentacremosa** [reply]
Hi @Jaimi5,

When testing the packet forwarding most of the time we got the following error below, but yesterday we were able to forward the packet once. 

What is this error -7? 

[3719655][V][LoraMesher.cpp:358] receivingRoutine(): [LoRaMesher] Stack space unused after entering the task: 2252
[3719666][V][LoraMesher.cpp:359] receivingRoutine(): [LoRaMesher] Free heap: 301512
[3719674][I][PacketService.cpp:11] createEmptyPacket(): [LoRaMesher] Packet created with 89 bytes
[3719683][I][LoraMesher.cpp:372] receivingRoutine(): [LoRaMesher] Receiving LoRa packet: Size: 89 bytes RSSI: -127 SNR: -7
[3719695][W][LoraMesher.cpp:382] receivingRoutine(): [LoRaMesher] Reading packet data gave error: -7
[3719854][V][LoraMesher.cpp:440] waitBeforeSend(): [LoRaMesher] Preamble detected while waiting 1
[3719863][V][LoraMesher.cpp:433] waitBeforeSend(): [LoRaMesher] RandomDelay 705 ms
[3720575][V][LoraMesher.cpp:734] printHeaderPacket(): [LoRaMesher] Packet send -- Size: 89 Src: 7934 Dst: 79D8 Id: 75 Type: 2 Via: 79D8 Seq_Id: 0 Num: 0

Thank you!

**Reply by polentacremosa** [reply]
And also, is there a way to induce packet forwarding via software, without making the sender node and the receiver one out of reach of each other?

**Reply by Jaimi5** [reply]
Hi! Sorry for the delay, the error you got is from RadioLib: [Error Codes](https://jgromes.github.io/RadioLib/group__status__codes.html). The -7 is a CRC mismatch and is a common error when you are sending messages with low power and there is some interferences.
To fix this I will recommend you setting the power to less than 6 (the default power) or decreasing the packet size you are sending.

**Reply by Jaimi5** [reply]
For the second part, yes, you can set a topology using software. Firstly you need to `#define LM_TESTING` in `BuildOption.h` and then, in loramesher.cpp, in the following function: 
`bool LoraMesher::canReceivePacket(uint16_t source) {
    return true;
}`
You can set the source whereas the node can receive from the source or not.
For example, if node 0x7934 can receive from node 0xABCD, you can just set this function:
```
bool LoraMesher::canReceivePacket(uint16_t source) {
    uint16_t local_addr = getLocalAddress();
    if (local_addr == 0x7934) {
       if (source == 0xABCD) return true; 
    } else if (local_addr == 0xABCD) {
       if (source == 0x7934) return true;
    }
    return false;
}
```

It is a little complicate because it was not intended to be used by other users. If you need more help I have some scripts that automate this proces given an adjacency graph.
Hope this helps. If you need more help we can schedule a meeting if you need, contact me via email directly!

**Reply by polentacremosa** [reply]
Hi,

Thank you so much for your response. 
I stopped to test and study other LoRa applications and will now test what you have suggested. 
I'm very grateful for your help and disponibility, I'll share updates of the test results soon.
